ITE/4/README.adoc

```diff
@@ -0,0 +1,462 @@
+
+| Sponsor
+| link:https://github.com/santiagotorres[Santiago Torres-Arias]
+
+| Status
+| Draft :speech_balloon:
+
+| Type
+| Standards
+
+| Created
+| 2019-09-30
+
+|===
+
+[[abstract]]
+== Abstract
+
+This ITE proposes a generic Unique Resource Identifier (URI) scheme for in-toto
+artifacts. Currently, the
+link:https://github.com/in-toto/docs/blob/master/in-toto-spec.md[in-toto specification]
+only allows for files in the local filesystem to be the artifacts of link
+metadata. This ITE defines the scheme for a generic URI and provides a
+description of the expected behaviour of in-toto implementations that support
+generic URIs.
+
+[[specification]]
+== Specification
+
+The current in-toto specification must be modified to provide support for
+generic URI schemes. The current specification allows only for files in the
+local filesystem to be set as artifacts. With the addition of other entities,
+the parts of the specification that directly interact or handle the artifacts
+must be updated.
+
+in-toto link metadata files currently look like this:
+
+```
+{
+    "_type": "link",
+    "name": "<NAME>",
+    "command": "<COMMAND>",
+    "materials": {
+        "<PATH>": <HASH>,
+        "...": ...
+    },
+    "products": {
+        "<PATH>": <HASH>,
+        "...": ...
+    },
+    "byproducts": {
+        "stderr": "<STDERR>",
+        "stdout": "<STDOUT>",
+        "return-value": "<RETURN-VALUE>"
+    },
+    "environment": {
+        "variables": "<ENV>",
+        "filesystem": "<FS>",
+        "workdir": "<PWD>"
+    }
+}
+```
+
+The `materials` and `products` fields contain key-value pairs. The URI for each
+artifact is its path in the filesystem. With this ITE active, the artifact could
+instead have a generic URI which can be resolved to some external resource that
+is part of the supply chain. The URI MUST have a fixed structure and MUST
+contain all the information necessary for the verifier to resolve if any part of
+the verification workflow requires it.
+
+This ITE proposes the following structure for the URI.
+
+`<scheme>:<hier-part>`
+
+This structure is derived from
+link:https://tools.ietf.org/html/rfc3986[RFC 3986]. The `scheme` defines the
+token or identifier for the type of resource the URI refers to. It can also
+contain other relevant information that indicates how the resource can be
+accessed. The `hier-part` identifies the path or location of the resource.
+
+A compliant in-toto system must be capable of generating link metadata with
+artifacts that have generic URI references. Therefore, this system must also be
+capable of resolving this URI and calculating cryptographic hashes of these
+artifacts. It's important to note that the contents of the resource are hashed,
+as opposed to the URI itself. This ITE doesn't specify how the contents are
+resolved and hashed by the system generating the metadata for the relevant
+abstract entities. All it specifies is that the resolution and the URI itself
```

Comment:
> This ITE doesn't specify how the contents are
> resolved and **hashed** by the system generating the metadata for the relevant
> abstract entities.

If you don't know how to hash something across different implementations, how can you verify items at URIs?  If I implement an in-toto verification routine for my package manager, how do I handle URIs I don't know?  Can different tools that generate in-toto link metadata choose to hash the same URI in different ways?  If so, this is concerning to me.

Reply:
While it can lead to divergence and problems with cross compatibility (this is explored in the Backward Compatibility section in the context of current implementations being able to verify metadata with generic URIs but perhaps a detailed analysis belongs in the specification section), I'm not sure we can enforce how these resources are hashed, since I believe it's very hard to generalise.

Reply:
ITE-4 is probably less specific than it seems through the motivating examples GitHub and SPDX. And I think there is merit in having a generic ITE like this. It wouldn't be feasible to define a generic way of hashing all the resources that might be added through this ITE.

But I also agree with @JustinCappos that it is important that all implementations that support a given scheme should (or rather must) follow the same specification for hashing the resource. @adityasaky, have you thought about how such a specification should be negotiated? Through separate ITEs? Or by the implementers?  Maybe ITE-4 could discuss this?

Reply:
> Through separate ITEs?

We could have an ITE for each but I'm wondering if we'd have a bunch as URI tokens are added. And these aren't changes to the core specification, so is it okay to use the ITE process like that?

Doing so also allows the in-toto community to assess and contribute to how abstract resources are hashed and recorded. I think a formal and open process, even if left to the implementers, is necessary to ensure security guarantees aren't lost. I also looked into having a single, growing ITE to collect schemes and hashing information, but that doesn't seem quite possible, as ITEs aren't designed for frequent change (which makes sense). Indeed, from ITE-1:

_"Ideally, all changes to an ITE should be completed before it is "Accepted", but surprises may still occur. Changes might be minor changes, such as clarifications or typo fixes, or major changes, which would alter the intent, scope, API, or core behavior of the ITE.

All changes to an already "Accepted" ITE, must be submitted via PR as usual. In the case of major changes, the Sponsor should also communicate those changes on the mailing list and take sufficient time to ensure there is consensus on the changes before merging them. A link to any discussion should be added to the PR for the change and to the References section. If there are significant objections or questions around proposed changes, the ITE Sponsor or Reviewer may choose to return the ITE to a "Draft" status for more extensive discussion and eventual review again for acceptance."_

Reply:
You are right. ITEs don't seem to be the right place to collect these resource definitions.

Reply:
Yep. Perhaps we can plan to formalise the _process_ of adding a token and how the relevant content should be hashed in a separate ITE, and mention that here?

cc @SantiagoTorres as well

Reply:
I believe that this is probably the best path forward, and it follows the philosophy that TUF has with their wireline formats ([TAP  11](https://github.com/theupdateframework/taps/blob/master/tap11.md)).

Reply:
Updated with:

> It is important to note that this ITE doesn’t mandate any specific generic URI tokens that all conforming implementations MUST support. Instead, it merely specifies what any generic URI MUST look like, and what rules the system MUST follow. However, it is vital to create a formal process that introduces new URI tokens and details how their abstract contents are hashed. This ensures that different in-toto implementations do not diverge significantly, and by creating an open process that involves the in-toto community, it can ensure consistency when hashing resources that multiple stakeholders use. Finally, such a process will also help ensure that the hashing of these entities is done in a correct and secure manner. However, detailing such a process is out of the scope of this ITE and must instead be handled in a future ITE.